### DISCUSSION TITLE
Usage with Non-JavaScript Clients

### DISCUSSION THREAD

**Discussion by nilskretschmer** [qruto/laravel-wave#55] [Q&A]
Hi,

thanks for this awesome package and your work on it!

Currently I investigate if I could use it within a project.

I have the following setup:
1. A central Laravel Backend Server ("The Control Server")
2. Several mobile devices each running their own web application (could also be Laravel)
3. The mobile devices can be in changing network environments and even behind (corporate) firewalls/routers
4. I need to establish a (bi-)directional communcation between the devices and the control server. The devices can run any backend/framework or language to connect to the control server. It is m2m communcation so no JavaScript Frontend or any user interaction.

I was thinking about using Websockets/SSE where each device initiates a connection against the Control Server which is publicly available. The problem is: How can I connect against the Laravel Backend - either using Websockets or SSE (using this package)? I would need some Laravel Echo for PHP or any other general purpose non-javascript language SDK. I think it should be possible using Guzzle using a long-running process/artisan command - but how to handle authentication/authorization?

Any thoughts on how to accomplish this?

**Comment by slavarazum** [comment]
Hi @nilskretschmer !

Laravel Echo or WaveJS are just clients to connect to your server. Since the SSE connection is simple HTTP request and considering that you need machine to machine communication you are be able to implement SSE connection with any HTTP client (like Guzzle). Authentication and authorization processes are also just a HTTP request params processing by Laravel server.
Using WebSockets will be a little bit more tricky, because you should deal with WebSocket type connections in addition to HTTP.

Let me know if you will have any troubles with implementation.

**Reply by nilskretschmer** [reply]
Hey @slavarazum 

thank you  for your response! I just implemented my own simple SSE controller that returns an event stream and I use `Guzzle` on the client to listen to those events. This works fine, but I would like to integrate `laravel-wave` / Laravel's Broadcasting into my application to use Laravel's "native" Broadcasting features (broadcasting events, channels etc..).

The problem:
I didn't find any up-to-date "client-side" PHP pusher implementation. That is also the same problem for any websocket solution (including the new Laravel Reverb which is also "only" server TO client). It seems like there are not that many use cases that need to have two different Laravel applications talk to each other (where one is a server and the other one is a client) in real time. I think this would be nice but currently I don't know where/how to start. I think the first thing would be to build some "client-side" pusher-compatible PHP SDK. Like Laravel Echo for PHP.

Do you have any idea how to solve that?

**Reply by slavarazum** [reply]
To establish the SSE connection with Wave, a good starting point would be to use the default [Laravel HTTP Client](https://laravel.com/docs/11.x/http-client) (not sure if it supports streamed response handling) or the raw GuzzleHttpClient. Emulate the appropriate request with the necessary headers:

```php
$headers = [
    'Cache-Control' => 'no-cache',
    'Pragma' => 'no-cache',
    'Connection' => 'keep-alive',
];
```

Authentication and authorization also require setting up appropriate request headers/parameters for your purposes. Next, you will deal with basic HTTP stream and Server-Sent Event format. Nothing related to Pusher here. I think it will be easier for you to setup http stream instead of a web-socket connection.

**Comment by nilskretschmer** [comment]
But I think Laravel Wave is Pusher compatible (because Laravel Broacasting is)? So to utilize these features I would need to implement some Pusher compatible PHP client to connect to the Laravel backend and listen to default Broadcasting Events and/or use presence/private channels.  Sure I can use Guzzle to just connect to an SSE endpoint. But that's what I already do without using wave or any default Laravel Broadcasting Events.

It's currently basically like:

```
class DeviceController extends Controller
{
    public function connect(ClientManager $manager)
    {
        ignore_user_abort(true);
        $client = $manager->getClientFromToken();
        $device = $client->device;
        
	$response = new StreamedResponse();
	$response->setCallback(function () use ($device){
            $event = array('event' => 'connect');
            echo json_encode($event) . PHP_EOL;
            $device->update(['attributes->status' => 'online']);

            while (true) {
                echo "; device: $device->id" . PHP_EOL;
                ob_flush();
                flush();
                if (connection_aborted()) {
                    $device->update(['attributes->status' => 'offline']);
                    break;
                }
                sleep(1);
            }
	});
		
	$response->headers->set('Content-Type', 'text/event-stream');
	$response->headers->set('X-Accel-Buffering', 'no');
	$response->headers->set('Cache-Control', 'no-cache');
	$response->send();
    }
}
```

And on the client-side:
```
<?php

namespace App\Console\Commands;

use App\Support\ControlServer;
use Illuminate\Console\Command;
use App\Models\Stream;

class Listen extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'control:listen';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Listens to control server';

    /**
     * Execute the console command.
     */
    public function handle(ControlServer $server)
    {
        try {
            $lines = $server->stream('/api/connect');
            foreach ($lines as $line) {
                if (strlen($line) > 0) {
                    $this->parse($line);
                }
            }
        } catch (\Exception $e) {
            $this->error($e->getMessage());
        }
    }
}
```

So no wave, no Laravel Events just SSE and Passport Client Credential Tokens for Auth.

**Reply by slavarazum** [reply]
It this example you created your own simple SSE connection endpoint.
If you want to use full power of Laravel Broadcasting with events, channels etc. I will recommend you to dig deeper into broadcasting documentation to correctly understand how exactly it works.
Pusher in this context is just protocol for WebSockets. Wave doesn't use WebSockets it implements SSE standard and just uses native `redis` broadcasting driver for [pub/sub](https://redis.io/docs/interact/pubsub/) paradigm.

**Reply by nilskretschmer** [reply]
You are right. There was a bit of a misunderstanding. My idea is more of having a laravel client library (I don't care if it would be for sse or websockets) that is able to listen to laravel event messages. So one laravel app would be the sender, the other app would be the listener/consumer. I think SSE would be easier than websockets. And yes for that we don't need pusher. But we would need to implement public, private and presence channels. I see that the laravel wave client is doing that as well.

It is like having laravel echo or the laravel wave client as a php client (for example a long running process or console command). I already looked into the laravel echo library. But it is quiet complex, because it uses pusher sdk behind the scenes. I also looked into the js wave client and I think building a client for php using guzzle would be easier, because the whole SSE concept is easier than websockets.

Do you think such a client would be possible? Do you see any problems? And could you explain in brief how the wave client handles authorization? I see that laravel's default broadcasting auth endpoint is called. But what happens then and how is authorization handled?

**Reply by slavarazum** [reply]
I'm pretty sure that you can create such PHP client. The trick is simply to send streamed HTTP request. 
Private and presence channels authentication is done using the default `/broadcasting/auth` endpoint. For more details on route registration, you can explore Laravel's [broadcasting sources](https://github.com/laravel/framework/blob/48246da2320c95a17bfae922d36264105a917906/src/Illuminate/Broadcasting/BroadcastManager.php#L75).

**Reply by nilskretschmer** [reply]
Thanks for your response @slavarazum !

I already experimented with the amphp/websocket-client and I already got a working connection. Now I need to handle the authentication/authorization part. I know about the `/broadcasting/auth` endpoint, but I am not sure how Laravel handles the authentication of channels afterwards and I didn't find details in the docs so I need to make a deep-dive into the source code of Laravel's Broadcasting and Event handling - especially how authentication/authorization is handled.

I think normally the session is used to authenticate a specific user. And after a successful call to `/broadcasting/auth` the user gets access to specific requested private/presence channels. But in the case of m2m communcation it should be some token-based authentication/authorization and I am not sure if Laravel's Broadcasting System is up to that.

**Reply by slavarazum** [reply]
As I can see, you have started using `amphp/websocket-client`.
Happy to help you with SSE integration when you return to Wave 😉
